crates/in_stream/src/tcp.rs

```diff
@@ -0,0 +1,323 @@
+            return Ok(());
+        }
+        let written = match self.stream.write(&self.write_buf) {
+            Ok(written) => written,
+            Err(e) if e.would_block() => return Ok(()),
+            Err(e) => return Err(e),
+        };
+        assert_eq!(written, self.write_buf.drain(..written).count());
+        Ok(())
+    }
+}
+
+impl InStream<&mut [u8], &[u8]> for InStreamTcp {
+    /// tcp streams should use urls like tcp://
+    const URL_SCHEME: &'static str = SCHEME;
+
+    fn raw_connect<C: InStreamConfig>(url: &Url2, config: C) -> Result<Self> {
+        let config = TcpConnectConfig::from_gen(config)?;
+        let addr = tcp_url_to_socket_addr(url)?;
+        let stream = net2::TcpBuilder::new_v4()?.to_tcp_stream()?;
+        stream.set_nonblocking(true)?;
+        match stream.connect(addr) {
+            Err(_) => Self::priv_new(
+                stream,
+                Some(TcpConnectingData {
+                    addr,
+                    connect_timeout: match config.connect_timeout_ms {
+                        None => None,
+                        Some(ms) => Some(
+                            std::time::Instant::now()
+                                .checked_add(std::time::Duration::from_millis(ms))
+                                .unwrap(),
+                        ),
+                    },
+                }),
+            ),
+            Ok(_) => Self::priv_new(stream, None),
+        }
+    }
+
+    fn read(&mut self, data: &mut [u8]) -> Result<usize> {
+        self.priv_process()?;
+        self.priv_write_pending()?;
+        if self.connecting.is_none() {
+            self.stream.read(data)
+        } else {
+            Err(Error::with_would_block())
+        }
+    }
+
+    fn write(&mut self, data: &[u8]) -> Result<usize> {
+        self.priv_process()?;
+        if self.connecting.is_none() {
+            if self.write_buf.is_empty() {
+                // in the 99% case we can just write without buffering
+                let written = match self.stream.write(data) {
+                    Ok(written) => written,
+                    Err(e) if e.would_block() => {
+                        self.write_buf.extend_from_slice(data);
+                        return Ok(data.len());
+                    }
+                    Err(e) => return Err(e),
+                };
+                if written < data.len() {
+                    self.write_buf.extend_from_slice(&data[written..]);
+                }
+                Ok(data.len())
+            } else {
+                // if we already have a buffer, append to it and
+                // try to write the whole thing
+                self.write_buf.extend_from_slice(data);
+                self.priv_write_pending()?;
+                Ok(data.len())
+            }
+        } else {
+            self.write_buf.extend_from_slice(data);
+            Ok(data.len())
+        }
+    }
+
+    fn flush(&mut self) -> Result<()> {
+        loop {
```

Comment:
I can't prove easily that this will terminate. What if the buffer refills at the start of each loop?

Reply:
I suppose it may not... will add a TODO for potential future timeout flush config